Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1918,66 +1937,134 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
-                                                          1, // Since this is glyph by glyph, glyphCount is one
-                                                          glyphRun->glyphIndices,
-                                                          nullptr,
-                                                          nullptr,
-                                                          glyphRun->isSideways,
-                                                          glyphRun->bidiLevel };
-                     D2D1_POINT_2F origin{ 0, 0 };
-                     // Iterate through every glyph by incrementing pointer in glyphIndices array
-                     for (uint32_t i = 0; i < glyphRun->glyphCount; ++i, ++(individualGlyphRun.glyphIndices)) {
-                         CGAffineTransform finalTextTransform = CGAffineTransformConcat(textTransform, runningGlobalTransform);
-                         deviceContext->SetTransform(__CGAffineTransformToD2D_F(finalTextTransform));
-                         deviceContext->DrawGlyphRun(origin, &individualGlyphRun, state.fillBrush.Get(), DWRITE_MEASURING_MODE_NATURAL);
-
-                         // Uses glyphAdvances to move each glyph
-                         runningGlobalTransform = CGAffineTransformTranslate(runningGlobalTransform, glyphRun->glyphAdvances[i], 0);
-                     }
-                     return S_OK;
-                 }));
+            Factory()->CreateTransformedGeometry(pathGeometry.Get(), __CGAffineTransformToD2D_F(finalTextTransform), &transformedGeometry));
+    }
+
+    HRESULT ret = Draw(_kCGCoordinateModeDeviceSpace, nullptr, [&](CGContextRef context, ID2D1DeviceContext* deviceContext) {
+        if (state.textDrawingMode & kCGTextFill) {
+            deviceContext->SetTransform(__CGAffineTransformToD2D_F(finalTextTransform));
+            D2D1_POINT_2F origin{ 0, 0 };
+            for (size_t i = 0; i < positionsVectors.size(); ++i) {
+                if (glyphRuns[i].attributes) {
+                    CGColorRef fontColor = (CGColorRef)CFDictionaryGetValue(glyphRuns[i].attributes, _kCGForegroundColorAttributeName);
+                    if (!fontColor) {
+                        CFBooleanRef useContextColor =
+                            (CFBooleanRef)CFDictionaryGetValue(glyphRuns[i].attributes, _kCGForegroundColorFromContextAttributeName);
+                        if (!useContextColor || !CFBooleanGetValue(useContextColor)) {
+                            // Neither given a color nor use current context color, so set the fill color to black
+                            CGContextSetRGBFillColor(context, 0, 0, 0, 1);
+                        }
+                    } else {
+                        CGContextSetFillColorWithColor(context, fontColor);
+                    }
+                }
+
+                DWRITE_GLYPH_RUN transformedGlyphRun{ glyphRuns[i].run->fontFace,   glyphRuns[i].run->fontEmSize,
```

Comment:
Even if we don't transform by glyph if we have > 1 runs we will need to do this to clip/stroke at once to prevent the shadowing/alpha issues.  I suppose we could make a fast path for "simple text transform w/ just fill" as that's going to be the most common scenario but I think that'd lead to a good amount of code duplication.